src/transform/evalAssert.ml

```diff
@@ -0,0 +1,133 @@
+        with
+          Not_found -> []
+      in
+
+      let rec instrument_instructions il s = match il with
+        | i1 :: i2 :: is ->
+          begin
+            match i1 with
+            | Set (lval, _, _, _)
+            | Call (Some lval, _, _, _, _) -> [i1] @ (make_assert (get_instrLoc i2) lval) @ instrument_instructions (i2 :: is) s
+            | _ -> i1 :: instrument_instructions (i2 :: is) s
+          end
+        | [i] ->
+          if List.length s.succs > 0 && (List.hd s.succs).preds |> List.length < 2 then begin
+            (* If the successor of this has more than one predecessor, it is a join point, and we can not query for the value there *)
+            let loc = get_stmtLoc (List.hd s.succs).skind in
+            match i with
+            | Set (lval, _, _, _)
+            | Call (Some lval, _, _, _, _) -> [i] @ (make_assert loc lval)
+            | _ -> [i]
+          end
+          else [i]
+        | _ -> []
+      in
+
+      let rec get_vars e =
+        match e with
+        | Lval (Var v, _) -> [v]
+        | UnOp (_, e, _) -> get_vars e
+        | BinOp (_, e1, e2, _) -> (get_vars e1) @ (get_vars e2)
+        | _ -> []
+      in
+
+      let instrument_join s =
+        match s.preds with
+        | [p1; p2] ->
+          (* exactly two predecessors -> join point, assert locals if they changed *)
+          (* Possible enhancement: It would be nice to only assert locals here that were modified in either branch *)
+          let join_loc = get_stmtLoc s.skind in
+          let asserts = List.map (fun x -> make_assert join_loc (Var x,NoOffset)) !locals |> List.concat in
+          self#queueInstr asserts; ()
+        | _ -> ()
+      in
+
+      let instrument_statement s =
+        instrument_join s;
+        match s.skind with
+        | Instr il ->
+          s.skind <- Instr (instrument_instructions il s);
+          s
+        | If (e, b1, b2, l,l2) ->
+          let vars = get_vars e in
+          let asserts loc vs = List.map (fun x -> make_assert loc (Var x,NoOffset)) vs |> List.concat in
+          let add_asserts block =
+            if List.length block.bstmts > 0 then
+              let with_asserts =
+                let b_loc = get_stmtLoc (List.hd block.bstmts).skind in
+                let b_assert_instr = asserts b_loc vars in
+                [cStmt "{ %I:asserts %S:b }" (fun n t -> makeVarinfo true "unknown" (TVoid [])) b_loc [("asserts", FI b_assert_instr); ("b", FS block.bstmts)]]
+              in
+              block.bstmts <- with_asserts
+            else
+              ()
+          in
+
+          add_asserts b1;
+          add_asserts b2;
+          s
+        | _ -> s
+      in
+      ChangeDoChildrenPost (s, instrument_statement)
+  end
+  let transform (ask:Cil.location -> Queries.ask) file = begin
+    visitCilFile (new visitor ask) file;
+    let assert_filename = global_replace (regexp "\\(.*\\)\\(/[^/]+\\)\\(/[^/]+\\.c\\)") "\\1/assert_transform\\3" file.fileName in
```

Comment:
Since we now always merge because of the `stdlib.c`, ` file.fileName` here is always `stdout`. (Spoken aside, why `stdout`?)

https://github.com/goblint/analyzer/blob/666f68dfc436556ce9aa448197d1794f04ad8a26/src/util/cilfacade.ml#L150-L153

So we would have to change how we compute where to put the output.

Reply:
Indeed, the merged `filename` shouldn't really be used for anything, but the filename from each function's, etc's location should be used instead because they refer to original names.

That would also mean actually producing the instrumented programs in an unmerged way: everything going to a file corresponding to one of the merged files. Although that might be infeasible to do, so it might just have to be that the annotated program is in its combined form and we are free to choose whatever filename we want for it, because there's no single right answer.

Reply:
This is now configurable via an option